cdap-ui/app/hydrator/controllers/create/toppanel-ctrl.js

```diff
@@ -234,35 +249,59 @@ class HydratorPlusPlusTopPanelCtrl{
     this.myPipelineApi.runPreview(params, pipelineConfig).$promise
       .then((res) => {
         this.previewStore.dispatch(
-          this.previewActions.setPreviewId(res.preview)
+          this.previewActions.setPreviewId(res.application)
         );
         let startTime = new Date();
         this.previewStartTime = startTime;
         this.startTimer();
         this.previewStore.dispatch(
           this.previewActions.setPreviewStartTime(startTime)
         );
-
-        this.startPollPreviewStatus(res.preview);
+        this.currentPreviewId = res.application;
+        this.startPollPreviewStatus(res.application);
       }, (err) => {
         this.previewLoading = false;
+        this.currentPreviewId = null;
         this.myAlertOnValium.show({
           type: 'danger',
           content: err.data
         });
       });
   }
 
+  stopPreview() {
+    if (!this.currentPreviewId) {
+      return;
+    }
+    let params = {
+      namespace: this.$state.params.namespace,
+      scope: this.$scope,
+      previewId: this.currentPreviewId
+    };
+    this.previewLoading = true;
+    this.stopTimer();
+    this.myPipelineApi
+        .stopPreview(params, {})
+        .$promise
+        .then(() => {
+          this.currentPreviewId = null;
+          this.previewLoading = false;
+          this.previewRunning = false;
+        });
```

Comment:
We might not be able to show all the time, say the user is running a preview and decides to go another state. We implicitly stop the preview and the response might not be immediately back